### PR TITLE
Proposal for nullable value handling

### PR DESCRIPTION
**Pull request by gdavison**
A proposal for handling of nullable int and float values. This should be extended for nullable booleans as well. After evaluation in the AWS Provider, it could be moved to the Plugin SDK.

`nullable.ValidateTypeStringNullableFloat()` replaces the existing `validateTypeStringNullableFloat()` in the AWS Provider. There is currently no Int equivalent in the AWS Provider, but there is one in the Kubernetes Provider (https://github.com/hashicorp/terraform-provider-kubernetes/blob/aae2fefb5bbe8fde1c2083d11ab2eaaaa63de657/kubernetes/validators.go#L204-L220).

Adds nullable value equivalents to `validation.<Type>Between()`, `validation.<Type>AtLeast()`, and `validation.<Type>AtMost()`. Could be extended with equivalents to `validation.IntDivisibleBy()`, `validation.IntInSlice()`, and `validation.IntNotInSlice()`.

Adds `nullable.TypeNullableInt` and `nullable.TypeNullableFloat` as aliases for `schema.TypeString` to make intention more clear and potentially guide linting, e.g. to require one of the appropriate validation functions.

Implements Go types to wrap the string representations of nullable values and adds functions `IsNull()` and `Value()`.

Using the type wrapper, we could replace the following code

```go
if v, ok := m["instance_warmup"]; ok {
	if s, ok := v.(string); ok && s != "" {
		i, _ := strconv.ParseInt(s, 10, 64)
		refreshPreferences.InstanceWarmup = aws.Int64(i)
	}
}
```

with

```go
if v, ok := m["instance_warmup"]; ok {
	if i, null, _ := nullable.Int(v.(string)).Value(); !null {
		refreshPreferences.InstanceWarmup = aws.Int64(i)
	}
}
```

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
NONE
```

Output from acceptance testing:

N/A

### CODE REVIEW

File: aws/internal/nullable/float.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	value, err := strconv.ParseFloat(string(f), 64)
+	if err != nil {
+		return 0, false, err
+	}
+	return value, false, nil
+}
+
+// ValidateTypeStringNullableFloat provides custom error messaging for TypeString floats
+// Some arguments require a floating-point value or unspecified, empty field.
+func ValidateTypeStringNullableFloat(v interface{}, k string) (ws []string, es []error) {
```
Comment (gdavison) [inline aws/internal/nullable/float.go:40]: I'm not sure if this full naming style is better, or the shorter names below. Shorter seems to fit better with Go conventions.

**Comment by gdavison** [comment]
In general, this needs a visit from the refactoring fairy, since there's a lot of duplication, but I'll wait for consensus on the approach before spending more time on this.

File: aws/internal/nullable/float.go (hunk 1 above)
Comment (jrhouston) [reply aws/internal/nullable/float.go:40]: +1 to shortening the name, it's a bit verbose at the moment.

**Comment by ewbankkit** [comment]
Relates:

* https://github.com/hashicorp/terraform-provider-aws/pull/19066

**Pull request closed without merging**